Kernel+Userland: Add jails

supercomputer7 opened the pull request:
This is something I talked about on in the discord server for a while, and the idea is quite simple - the goal is to perform process isolation in a simple way that can be adapted for non-native programs of the SerenityOS project, (i.e. ports), to provide a full solution together with the pledge+unveil mechanism we already have and to cover whatever that doesn't use the latter method.

Suggestion are welcome, as always :)

tslater2006 commented on Kernel/Process.h:
```diff
@@ -248,11 +253,11 @@ class Process final
-    static void for_each_in_pgrp(ProcessGroupID, Callback);
-    template<IteratorFunction<Process&> Callback>
-    void for_each_child(Callback);
+    static void for_globally_each(Callback);
```
I think naming this `for_each_globally` would read better and match the naming pattern of the other for_each_* methods below.

supercomputer7 commented:
I should probably update the sysfs document about the new `jails` node, as well as to write manual pages for the new utilities.

ADKaster commented on Kernel/Arch/x86/init.cpp:
```diff
@@ -237,6 +238,7 @@ extern "C" [[noreturn]] UNMAP_AFTER_INIT void init(BootInfo const& boot_info)
     __stack_chk_guard = get_fast_random<size_t>();
 
     ProcFSComponentRegistry::initialize();
+    JailManagement::the();
```
Does this need added to aarch64 init too?

ADKaster commented on Kernel/Jail.h:
```diff
@@ -0,0 +1,48 @@
+
+class JailManagement;
+
+AK_TYPEDEF_DISTINCT_ORDERED_ID(u64, JailIndex);
```
If JailIndex is `u64`, then why does the syscall take `int`? that's not `u64` on any platform.

ADKaster commented on Kernel/Jail.h:
```diff
@@ -0,0 +1,48 @@
+
+class Jail : public RefCounted<Jail> {
+    friend class JailManagement;
+    friend class Process;
```
Why does process need to be this guy's friend?

ADKaster commented on Kernel/Process.cpp:
```diff
@@ -63,6 +63,67 @@ SpinlockProtected<Process::List>& Process::all_instances()
+ErrorOr<void> Process::for_each_with_jail_respect(Function<ErrorOr<void>(Process&)> callback)
+{
+    ErrorOr<void> result {};
+    jail().with([&](auto my_jail) {
+        Process::all_instances().with([&](auto const& list) {
+            for (auto it = list.begin(); it != list.end();) {
+                auto& process = *it;
+                ++it;
+                process.jail().with([&](auto& their_jail) {
+                    if (!my_jail || their_jail.ptr() == my_jail.ptr())
+                        result = callback(process);
+                });
+                if (result.is_error())
+                    break;
+            }
+        });
+    });
```
This dance makes me nervous. We take the spinlocks in the order Process::jail --> Process::all_instances --> Process::jail. Is this always safe?

Is it also looks like it will lock `this`'s jail spinlock twice once we get to it in the all processes list?

I'm not convinced that iterating this list requires manual `it` varaibles the way you're doing it. does ranged for work?

Do we really want to call the callback while holding all 3 spin locks?

I wonder if it would be ok to aggregate a list of processes in a Vector with some inline capacity here on the stack and call the callback after gathering them under the lock?

supercomputer7 replied:
I changed it a bit now and it should take a different path to this. Also, if we keep taking our jail spinlock and the second process spinlock under the spinlock of `Process::all_instances()`, in theory this should prevent a deadlock from happening because no two processes can compete over each other jail spinlock unless they somehow both acquire the `Process::all_instances` which should never happen...

ADKaster commented on Kernel/Process.cpp:
```diff
@@ -63,6 +63,67 @@ SpinlockProtected<Process::List>& Process::all_instances()
+                auto& process = *it;
+                ++it;
+                process.jail().with([&](auto& their_jail) {
+                    if ((!my_jail || their_jail.ptr() == my_jail.ptr()) && (process.ppid() == my_pid || process.has_tracee_thread(pid())))
```
here and elsewhere, we don't need to take another process's jail lock to know that we don't have a jail. I feel like even though it may be duplicated, it will be useful to split this out into two branches where we don't bother taking the 3rd level of lock when we don't have our own jail.

ADKaster commented on Kernel/Process.cpp:
```diff
@@ -657,6 +718,18 @@ void Process::finalize()
     m_executable.with([](auto& executable) { executable = nullptr; });
+    m_attached_jail.with([](auto& jail) {
+        if (jail) {
+            jail->attach_count().with([&](auto& attach_count) {
```
can we make this entire logic here a member function on Jail?

`Jail::detach()`

Then we don't need to touch its private members from Process.

ADKaster commented on Kernel/Process.h:
```diff
@@ -106,6 +108,7 @@ struct LoadResult;
 class Process final
     : public ListedRefCounted<Process, LockType::Spinlock>
     , public LockWeakable<Process> {
+    friend class Jail;
```
Why does Jail need to be a friend of process?

ADKaster commented on Kernel/Process.h:
```diff
@@ -248,11 +253,11 @@ class Process final
-    void for_each_child(Callback);
+    static void for_each_ignoring_jails(Callback);
+
+    ErrorOr<void> for_each_with_jail_respect(Function<ErrorOr<void>(Process&)>);
```
should `for_each_with_jail_respect` be static?

supercomputer7 replied:
I rather use it with `Process::current()` prefix to indicate it relies on it.

supercomputer7 replied:
On second thought, we could just put it inside the method and it will almost look the same, so let's do that...

ADKaster commented on Kernel/Process.h:
```diff
@@ -248,11 +253,11 @@ class Process final
+
+    ErrorOr<void> for_each_with_jail_respect(Function<ErrorOr<void>(Process&)>);
+    ErrorOr<void> for_each_in_pgrp_with_jail_respect(ProcessGroupID, Function<ErrorOr<void>(Process&)>);
+    ErrorOr<void> for_each_child_with_jail_respect(Function<ErrorOr<void>(Process&)>);
```
I'm not sure I'm sold on the `with_jail_respect` suffix here from a grammar point of view. maybe `for_each_*_respecting_jails` ? `jail_aware_for_each` ? `for_each_*_in_same_jail` ?

ADKaster commented on Kernel/Syscalls/jail.cpp:
```diff
@@ -0,0 +1,58 @@
+    return 0;
+}
+
+ErrorOr<FlatPtr> Process::sys$jail_attach(Userspace<Syscall::SC_jail_attach_params const*> user_params)
```
This interface requires a process to add itself to a jail initially. So, this is something that must be done just after fork(), before exec? What about in the posix_spawn case? 

Should we add a non-portable posix_spawnattr item to add the new process to a jail? 

What about adding a new process to a jail externally? Should it be possible for a process to send another process to jail?

ADKaster replied:
In this case, I'm thinking about a server like `dockerd` that will manage a set of applications each in their own jail that might like to live itself in a non-jailed state but posix_spawn off some processes with a jail, and occasionally run commands in the context of that jail.

supercomputer7 replied:
Look at the `jail-attach` utility - it doesn't do any sort of `fork()` there and still enters into the a jail. You could also enter a jail and then do fork+exec and you still will be in a jail, and it also works the same if you join a jail after `fork()` but before `exec()`.

ADKaster commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -979,6 +979,21 @@ ErrorOr<void> adjtime(const struct timeval* delta, struct timeval* old_delta)
 #endif
 
+#ifdef AK_OS_SERENITY
+ErrorOr<void> join_jail(int jail_index)
```
Typically we add these things to both LibC/serenity.cpp and LibCore/System.cpp. I'm not sure if we need to keep that pattern up now that all our own apps are using LibCore, but if someone ports a FreeBSD program that knows how to use jails it might be nice to have a C API.

supercomputer7 replied:
To be honest, I am not aiming for feature-parity or even compatibility with FreeBSD jails in the long run, so I just neglected to add functions nobody will use into LibC.

ADKaster commented on Userland/Libraries/LibCore/System.h:
```diff
@@ -167,6 +167,11 @@ enum class SearchInPath {
 
+#ifdef AK_OS_SERENITY
+ErrorOr<void> join_jail(int jail_index);
+ErrorOr<void> create_jail(StringView jail_name);
```
`ErrorOr<void>` ?  Doesn't the jail_create syscall return the jail id?

ADKaster commented on Userland/Utilities/jail-attach.cpp:
```diff
@@ -0,0 +1,52 @@
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    int jail_index = 0;
```
the jail indices are u64 in the kernel, so if you make this one as well, that's one less error check below.

ADKaster commented on Userland/Utilities/jail-attach.cpp:
```diff
@@ -0,0 +1,52 @@
+            continue;
+
+        // FIXME: Allow a custom selection of variables once ArgsParser supports options with optional parameters.
+        if (!preserve_env && env_view.substring_view(0, maybe_needle.value()) != "TERM"sv)
```
This looks suspiciously copied from `pls` and `Escalator`. Maybe we should move it into a common LibCore function?

ADKaster reviewed (commented):
I saw it wasn't a draft anymore, so here's some comments :)

LucasChollet commented on Userland/Utilities/jail-attach.cpp:
```diff
@@ -0,0 +1,27 @@
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    unsigned jail_index = 0;
```
Can't we pledge an empty set here?
Totally outside my scope, but we might want to add a `jail` pledge.

supercomputer7 replied:
There's already a `jail` pledge I implemented, so I should use it here, thanks for noticing :)

LucasChollet commented on Userland/Utilities/jail-create.cpp:
```diff
@@ -0,0 +1,25 @@
+
+ErrorOr<int> serenity_main(Main::Arguments arguments)
+{
+    String new_jail_name;
```
Same here

LucasChollet replied:
A `StringView` will be preferred here

LucasChollet commented on Userland/Utilities/lsjails.cpp:
```diff
@@ -0,0 +1,35 @@
+    TRY(Core::System::unveil("/sys/kernel/jails", "r"));
+    TRY(Core::System::unveil(nullptr, nullptr));
+
+    auto jails_data = TRY(Core::File::open("/sys/kernel/jails", Core::OpenMode::ReadOnly));
```
`Core::File` is deprecated for reading and writing, please use `Core::Stream::File` instead

LucasChollet reviewed (commented):
Some userland comments

supercomputer7 commented:
> Some userland comments

Thanks for the review! I addressed the issues according to your comments. I also wrote manual pages for the new 3 utilities.
If @ADKaster doesn't have anything left to suggest, I think this is ready for a merge :)

supercomputer7 commented:
I force-pushed a quick fix due to a bug in the attach count of a Jail - instead of doing an increment on it during `fork` and on the `jail_attach` syscalls, I did the increment in the `fork` and `execve` syscalls, which under a quick test in `/bin/Shell` inside a Jail, led to leaking the attach count so the Jail could never been cleaned up when the last process exited from it.
Now it should be good to go.
